File: rust/src/problems/p146.rs
use std::collections::HashMap;

struct Node {
    key: i32,
    value: i32,
    prev: *mut Node,
    next: *mut Node,
}

pub struct LRUCache {
    dummy: Box<Node>,
    map: HashMap<i32, Box<Node>>,
    capacity: usize,
}

impl LRUCache {
    pub fn new(capacity: i32) -> Self {
        let mut dummy = Box::new(Node {
            key: -1,
            value: -1,
            prev: std::ptr::null_mut(),
            next: std::ptr::null_mut(),
        });
        dummy.next = dummy.as_mut();
        dummy.prev = dummy.as_mut();
        Self {
            dummy,
            map: HashMap::new(),
            capacity: capacity as usize,
        }
    }

    pub fn get(&mut self, key: i32) -> i32 {
        if let Some(node) = self.map.get_mut(&key) {
            unsafe {
                let node = node.as_mut();
                // is head
                if (*node.prev).key == -1 {
                    return node.value;
                }

                // remove node
                (*node.prev).next = node.next;
                (*node.next).prev = node.prev;

                // insert head
                let head = self.dummy.next;
                self.dummy.next = node;
                node.prev = self.dummy.as_mut();
                (*head).prev = node;
                node.next = head;

                node.value
            }
        } else {
            -1
        }
    }

    pub fn put(&mut self, key: i32, value: i32) {
        if let Some(node) = self.map.get_mut(&key) {
            unsafe {
                node.value = value;
                // is head
                if (*node.prev).key == -1 {
                    return;
                }

                // remove node
                (*node.prev).next = node.next;
                (*node.next).prev = node.prev;

                // insert head
                let head = self.dummy.next;
                self.dummy.next = node.as_mut();
                node.prev = self.dummy.as_mut();
                (*head).prev = node.as_mut();
                node.next = head;
            }
        } else {
            if self.map.len() == self.capacity {
                let tail = self.dummy.prev;
                unsafe {
                    (*(*tail).prev).next = self.dummy.as_mut();
                    self.dummy.prev = (*tail).prev;
                    self.map.remove(&(*tail).key);
                }
            }
            let mut node = Box::new(Node {
                key,
                value,
                prev: self.dummy.as_mut(),
                next: self.dummy.next,
            });
            let n: *mut Node = node.as_mut();
            self.map.insert(key, node);
            unsafe {
                (*self.dummy.next).prev = n;
                self.dummy.next = n;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let mut lru = LRUCache::new(2);
        lru.put(1, 1);
        lru.put(2, 2);
        assert_eq!(1, lru.get(1));
        lru.put(3, 3);
        assert_eq!(-1, lru.get(2));
        lru.put(4, 4);
        assert_eq!(-1, lru.get(1));
        assert_eq!(3, lru.get(3));
        assert_eq!(4, lru.get(4));
    }

    #[test]
    fn case2() {
        let mut lru = LRUCache::new(2);
        lru.put(1, 0);
        lru.put(2, 2);
        assert_eq!(0, lru.get(1));
        lru.put(3, 3);
        assert_eq!(-1, lru.get(2));
        lru.put(4, 4);
        assert_eq!(-1, lru.get(1));
        assert_eq!(3, lru.get(3));
        assert_eq!(4, lru.get(4));
    }

    #[test]
    fn case3() {
        let mut lru = LRUCache::new(2);
        lru.put(2, 1);
        lru.put(1, 1);
        lru.put(2, 3);
        lru.put(4, 1);
        assert_eq!(-1, lru.get(1));
        assert_eq!(3, lru.get(2));
    }
}

File: cpp/p146.cpp
#include <cassert>
#include <unordered_map>

using namespace std;

struct Node {
    int key;
    int val;
    Node *prev;
    Node *next;
    Node() {}
    Node(int key, int val) : key(key), val(val) {}
};

class LRUCache {
    int capacity;
    Node dummy;
    unordered_map<int, Node *> nodes;

    void append_to_tail(Node *n) {
        auto tail = dummy.prev;
        tail->next = n;
        n->prev = tail;
        n->next = &dummy;
        dummy.prev = n;
    }

    Node *get_node(int key) {
        if (nodes.count(key) == 0) {
            return nullptr;
        }
        Node *n = nodes[key];
        n->prev->next = n->next;
        n->next->prev = n->prev;
        append_to_tail(n);
        return n;
    }

public:
    LRUCache(int capacity) : capacity(capacity) {
        dummy.prev = &dummy;
    }

    int get(int key) {
        auto n = get_node(key);
        return n ? n->val : -1;
    }

    void put(int key, int value) {
        auto n = get_node(key);
        if (n) {
            n->val = value;
        } else {
            if (nodes.size() == capacity) {
                n = dummy.next;
                dummy.next = n->next;
                n->next->prev = &dummy;
                nodes.erase(n->key);
                delete n;
            }
            n = new Node(key, value);
            nodes.insert({key, n});
            append_to_tail(n);
        }
    }
};

int main() {
    auto cache = LRUCache(2);
    cache.put(1, 1);
    cache.put(2, 2);
    assert(1 == cache.get(1));
    cache.put(3, 3);
    assert(-1 == cache.get(2));
    cache.put(4, 4);
    assert(-1 == cache.get(1));
    assert(3 == cache.get(3));
    assert(4 == cache.get(4));
    return 0;
}